lib/filesystem/src/filesystem.cpp

```diff
@@ -546,6 +536,58 @@ namespace irods::experimental::filesystem::NAMESPACE_IMPL
         return p1_info.id == p2_info.id;
     }
 
+    auto data_object_size(rxComm& _comm, const path& _p) -> std::uintmax_t
+    {
+        if (_p.empty()) {
+            throw filesystem_error{"empty path", make_error_code(SYS_INVALID_INPUT_PARAM)};
+        }
+
+        detail::throw_if_path_length_exceeds_limit(_p);
+
+        if (!is_data_object(_comm, _p)) {
+            throw filesystem_error{"path does not point to a data object", _p, make_error_code(SYS_INVALID_INPUT_PARAM)};
+        }
+
+        // TODO Determine whether "DATA_REPL_NUM" is needed here.
+        const auto gql = fmt::format("select DATA_SIZE, DATA_REPL_NUM, DATA_MODIFY_TIME "
+                                     "where"
+                                     " COLL_NAME = '{}' and"
+                                     " DATA_NAME = '{}' and"
+                                     " DATA_REPL_STATUS = '1'",
+                                     _p.parent_path().c_str(),
+                                     _p.object_name().c_str());
+
+        irods::experimental::query_builder qb;
+
+        if (const auto zone = zone_name(_p); zone) {
+            qb.zone_hint(*zone);
+        }
+
+        auto query = qb.build(_comm, gql);
+
+        if (query.size() == 0) {
+            throw filesystem_error{"no good replica found", _p, make_error_code(SYS_NO_GOOD_REPLICA)};
+        }
+
+        // This implementation assumes that any good replica will always satisfy
+        // the requirement within the loop, therefore the first iteration always causes
+        // the size to be captured. The size object should be empty if and only if there
+        // are no good replicas.
+        std::uint64_t latest_mtime = 0;
+        std::uintmax_t size = 0;
```

Comment:
this probably needs to be wrapped in a `std::optional` because the query could result in no good replicas.

Reply:
instead of using `std::optional`, the query object should be checked to see if any results exist and act appropriately.